### PR TITLE
with-class should work with form-2 components

### PR DESCRIPTION
**Pull request by smogg**
with this implementation, calling `[with-class sorted-table ks rows state]` will correctly render the component. It will also work with standard hiccups, tho the returned value is no longer as simple, so this won't evaluate to true:

```clojure
(is (= (with-class "some-class" [:a])
       [:a {:class "some-class"}]))
```
I think that `with-class` is not the best idea in the first place - components we create should accept attributes by default, without need for additional fn. 
But since it's here and we find use for it, `with-class` should work in a way that:

- works with hiccups and doesn't complicate the output
- works with form-2 components and properly renders them usin `[]`, so they are reactive

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -8,7 +8,7 @@
                         ["snapshots" :clojars]]
 
   :dependencies [[org.clojure/clojurescript "1.8.51" :scope "provided"]
-                 [reagent "0.5.1"]
+                 [reagent "0.6.0-alpha"]
```
Comment (lvh) [inline project.clj]: Why is this changed in this PR?

File: project.clj [hunk 2]
```diff
@@ -42,7 +42,7 @@
                                          :pretty-print true}}}}
 
   :figwheel {:http-server-root "public"
-             :server-port 3449
+             :server-port 3001
```
Comment (lvh) [inline project.clj]: Why is this changed in this PR?

Comment (smogg) [inline project.clj]: checked in `project.clj` by mistake, my bad

File: src/wilson/dom.cljs [hunk 3]
```diff
@@ -174,9 +174,15 @@
   (update elem 1 merge-attrs attrs))
 
 (defn with-class
-  "Adds a class to an element."
-  [cls elem]
-  (with-attrs {:class cls} elem))
+  "Adds a class to a component."
+  [cls component]
+  (if (vector? component)
```
Comment (lvh) [inline src/wilson/dom.cljs]: You don't care if this is a vector, you care if it's a vector where the first element is a keyword.

The easiest way to do this is with core.match. Then you can also easily handle the case for `[:a]`, `[:a {:class "whatever"}]`, `[:a {:id "whatever"}]`, `[:a [:div] [:div]]`...

Comment (smogg) [inline src/wilson/dom.cljs]: Technically I care if it's a fn in the second case, cause then what I do is completely different. I don't see how match helps me here? If it's a vector (with or without map of attrs) then it's a hiccup, right? I don't need to know more

Comment (lvh) [inline src/wilson/dom.cljs]: I think we're thinking of different call signatures, I'm trying to make sure that `(with-class "yolo" [thingy-component a b c])` works; so "vector" doesn't tell you much?

Comment (smogg) [inline src/wilson/dom.cljs]: I see. Right now you can call it like this: `[with-class "yolo" thingy-component a b c]`. But I presume _forcing_ user to only use `()` makes a lot of sense

Comment (lvh) [inline src/wilson/dom.cljs]: Yeah, I dislike having two drastically different ways you use something even if it has to do something drastically different on the inside

**Comment by smogg** [comment]
Ok, so now `with-class` should be used _only_ with `()` and it works like this with form-2:

`(with-class "yolo" [form-2-component arg1 arg2])`

I agree this is better (and probably more intuitive) and I think we are ready to close #16

File: src/wilson/dom.cljs [hunk 4]
```diff
@@ -174,9 +175,19 @@
...
+  "Adds a class to a component."
+  [cls component]
+  (let [parse-f2c (fn [render-fn & outer-props]
+                    (let [outer-comp (apply render-fn outer-props)]
+                      (fn [_ & inner-props]
+                        (let [inner-comp (apply outer-comp inner-props)]
+                          (with-attrs {:class cls} inner-comp)))))
+        [render-fn] (match component
+                      [render-fn & _]
+                      [render-fn])]
```
Comment (lvh) [inline src/wilson/dom.cljs]: match does not do anything useful here! you could just use `first`. If you wanted to use match appropriately, you should use [core.match with a guard][guard]. This would have the benefit that it will be easier to handle the cases with and without an attr map already present. Maybe this does that already; you should have a test for that :)

[guard]: https://github.com/clojure/core.match/wiki/Basic-usage#guards

File: test/wilson/dom_test.cljs [hunk 5]
```diff
@@ -268,7 +268,18 @@
...
+        rows [{:a 1 :b "B" :c -6}
+              {:a 4 :b "A" :c 2}
+              {:a 3 :b "D" :c 4}]
+        ks (d/prepare-keys [:a :b :c])]
+    (with-mounted-component (d/with-class "some-class" [d/sorted-table ks rows state])
+      (fn [c div]
+        (let [table-el (.querySelector div "table")
+              has-class? (fn [cls el] (.contains (.-classList el) cls))]
+          (testing "with-class on form-2 component"
+            (is (has-class? "some-class" table-el))))))))
```
Comment (lvh) [inline test/wilson/dom_test.cljs]: I really like this test, although it does imply only working when running the tests on the browser... I guess that's unavoidable for a library like wilson.

File: src/wilson/dom.cljs (hunk 4 above)
Comment (smogg) [inline src/wilson/dom.cljs]: Uh, right. 
It works on elements without attrs map already and tests are there (first `with-class-test` doesn't have attrs-map, same in `with-attrs-test`)

File: src/wilson/dom.cljs [hunk 6]
```diff
@@ -174,9 +174,16 @@
...
-  [cls elem]
-  (with-attrs {:class cls} elem))
+  "Adds a class to a component."
+  [cls component]
+  (let [parse-f2c (fn [render-fn & outer-props]
+                    (let [outer-comp (apply render-fn outer-props)]
+                      (fn [_ & inner-props]
+                        (let [inner-comp (apply outer-comp inner-props)]
+                          (with-attrs {:class cls} inner-comp)))))]
+    (if (keyword? (first component))
```
Comment (lvh) [inline src/wilson/dom.cljs]: OK, cool :) You can write it either with or without match; match would have the benefit that you don't create parse-f2c when it's not necessary. You could also alternatively do that by making parse-f2c an external fn, and use `partial` to pass `cls` to it.

Comment (lvh) [inline src/wilson/dom.cljs]: (I'm not sure what the performance consequences of this are, though...)

File: test/wilson/dom_test.cljs (hunk 5 above)
Comment (smogg) [inline test/wilson/dom_test.cljs]: In what case would it work in the browser but not _elsewhere_? How else would you like me to test it?

File: src/wilson/dom.cljs (hunk 6 above)
Comment (smogg) [inline src/wilson/dom.cljs]: I think `match` might be an overkill for such simple purpose and I'd leave it as is

File: test/wilson/dom_test.cljs (hunk 5 above)
Comment (lvh) [inline test/wilson/dom_test.cljs]: `querySelector`, `.-classList`, ... imply a DOM, which you're only going to have in a browser. An additional test (note: not alternative) that would let you verify that this does what you think it does would be to pretend to be that part of reagent, and look more like the test for a hiccup form: assert that you get a vector back with some weird component in the first place, then the arguments you were expecting. Then render the component (well, call it ;)), and see that you get the appropriate hiccup structure back. All hiccup, no DOM :)
